provider/server/server_suite_test.go

```diff
@@ -38,38 +47,97 @@ func TestServer(t *testing.T) {
-	})
-	Expect(err).NotTo(HaveOccurred())
+	logf.SetLogger(GinkgoLogr)
 
-	machineClasses, err := mcr.NewMachineClassRegistry([]iri.MachineClass{
+	machineClasses := []iriv1alpha1.MachineClass{
 		{
-			//TODO
 			Name: machineClassx3xlarge,
-			Capabilities: &iri.MachineClassCapabilities{
+			Capabilities: &iriv1alpha1.MachineClassCapabilities{
 				CpuMillis:   4000,
 				MemoryBytes: 8589934592,
 			},
 		},
-	})
+	}
+	machineClassData, err := json.Marshal(machineClasses)
+	Expect(err).NotTo(HaveOccurred())
+
+	machineClassesFile, err := os.CreateTemp(GinkgoT().TempDir(), "machineclasses")
 	Expect(err).NotTo(HaveOccurred())
+	defer func() {
+		_ = machineClassesFile.Close()
+	}()
+	Expect(os.WriteFile(machineClassesFile.Name(), machineClassData, 0666)).To(Succeed())
 
-	srv, err = server.New(server.Options{
-		BaseURL:        baseURL,
-		MachineStore:   machineStore,
-		MachineClasses: machineClasses,
-	})
+	By("starting the app")
+
+	userHomeDir, err := os.UserHomeDir()
 	Expect(err).NotTo(HaveOccurred())
+
+	fs := pflag.NewFlagSet("test", pflag.ContinueOnError)
+	apinetPlugin := networkinterfaceplugin.NewDefaultOptions()
+	apinetPlugin.PluginName = "apinet"
+	apinetPlugin.AddFlags(fs)
+	Expect(fs.Set("apinet-node-name", "test-node")).NotTo(HaveOccurred())
+
+	opts := app.Options{
+		//TODO: set these using env variables if needed
+		Address:                     fmt.Sprintf("%s/test.sock", os.Getenv("PWD")),
+		BaseURL:                     baseURL,
+		PathSupportedMachineClasses: machineClassesFile.Name(),
+		RootDir:                     fmt.Sprintf("%s/libvirt-provider", userHomeDir),
+		StreamingAddress:            streamingAddress,
+		Libvirt: app.LibvirtOptions{
+			Socket:                "/var/run/libvirt/libvirt-sock",
+			URI:                   "qemu:///system",
+			PreferredDomainTypes:  []string{"kvm", "qemu"},
+			PreferredMachineTypes: []string{"pc-q35", "pc-i440fx-2.9", "pc-i440fx-2.8"},
+			Qcow2Type:             "exec",
+		},
+		NicPlugin: apinetPlugin,
+		// TODO: add other required fields
+	}
+
+	srvCtx, cancel := context.WithCancel(context.Background())
+	DeferCleanup(cancel)
+
+	go func() {
+		defer GinkgoRecover()
+		Expect(app.Run(srvCtx, opts)).To(Succeed())
+	}()
+
+	Eventually(func() (bool, error) {
+		return isSocketAvailable(opts.Address)
+	}, "30s", "500ms").Should(BeTrue(), "The UNIX socket file should be available")
+
+	address, err := machine.GetAddressWithTimeout(3*time.Second, fmt.Sprintf("unix://%s", opts.Address))
+	Expect(err).NotTo(HaveOccurred())
+
+	gconn, err := grpc.Dial(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	Expect(err).NotTo(HaveOccurred())
+
+	machineClient = iriv1alpha1.NewMachineRuntimeClient(gconn)
+	DeferCleanup(gconn.Close)
+
+	c := dialers.NewLocal()
+	libvirtConn = libvirt.NewWithDialer(c)
+	err = libvirtConn.Connect()
+	Expect(err).NotTo(HaveOccurred())
+	Expect(libvirtConn.IsConnected(), BeTrue())
+
+	DeferCleanup(libvirtConn.ConnectClose)
```

Comment:
I would suggest doing all the cleanup in the AfterSuite block, in place of using DeferCleanup in multiple places.

Reply:
`DeferCleanup` has it's purpose if you e.g. are creating multiple `Machine`s and want to clean up ASAP since the GH runner instance has only limited capacity. Also there is no real need for an `AfterSuite` if you `DefereCleanup` where things are being defined: being it global clients in the suite setup or `It` block related resources.

Reply:
ok, in that case DefereCleanup can be used for these purposes. However, I recommend having AfterSuite, where overall test artifacts cleanup has to be done.

Reply:
Minimally i don't like how is defer DeferCleanup used. It completely ignoring errors returned from functions, here it isn't big problem but i think we have to check if machine deletion was successful.

Reply:
`DeferCleanup` does actually have proper error handling - so if the cleanup fails, your test will fail. Regarding the general usage of `DeferCleanup`: If you do the `Object` cleanup in an `AfterSuite` or `AfterEach` run, you need to keep track of when you created what which is a tedious task and you might forget something along the way leading to weird testing behaviours.

Reply:
The `ginkgo` community has also a take on this topic: https://onsi.github.io/ginkgo/#cleaning-up-our-cleanup-code-defercleanup

> Setup and cleanup patterns like the one above are common in Ginkgo suites. While powerful, however, AfterEach nodes have a tendency to separate cleanup code from setup code. We had to create an originalWeightUnits closure variable to keep track of the original environment variable in the BeforeEach and pass it to the AfterEach - this feels noisy and potential error-prone.

Reply:
Below can be the decision factors:
- **Resource Scope**: Consider whether cleanup is specific to individual tests or should be performed globally after all tests.
- **Isolation**: If tests need to be isolated and not share state, DeferCleanup might be more suitable.
- **Efficiency**: If cleanup can be consolidated and performed more efficiently once after all tests, AfterSuite might be preferred.

Reply:
@afritzler  DeferCleanup have proper handling of error but i don't see it in our code. You have to implement proper handlig of error with Expect() functions which are missing.

Reply:
DeferCleanup are fixed

Reply:
Addressed.